File: leetcode/c/remove-linked-list-elements.c
/*
Remove all elements from a linked list of integers that have value val.

Example:

Input:  1->2->6->3->4->5->6, val = 6
Output: 1->2->3->4->5
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* removeElements(struct ListNode* head, int val) {
    struct ListNode *p, *temp, *pp, *new_head;
    
    if (!head)
        return head;
    new_head = head;
    p = new_head;
    pp = p;
    while (p) {
        if (p->val == val) {
            if (p == new_head) {
                temp = p;
                p = p->next;
                new_head = p;
            } else {
                temp = p;
                p = p->next;
                pp->next = p;
            }
            free(temp);
        } else {
            pp = p;
            p = p->next;        
        }
    }
    return new_head;
}

File: leetcode/go/link-list/remove-linked-list-elements.go
/*
Remove all elements from a linked list of integers that have value val.

Example:

Input:  1->2->6->3->4->5->6, val = 6
Output: 1->2->3->4->5
*/
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeElements(head *ListNode, val int) *ListNode {
    var p, pp, new_head *ListNode
    
    if head == nil {
        return head
    }
    new_head = head
    p = new_head
    pp = p
    for ; p != nil; {
        if p.Val == val {
            // remove element p          
            if p == new_head {
                // p is at head of list
                p = p.Next
                new_head = p
            } else {
                p = p.Next
                pp.Next = p
            }
        } else {
            pp = p
            p = p.Next
        }
    }
    return new_head
}
